pkg/shell/process.go

```diff
@@ -74,16 +74,61 @@ func (p *Process) OnStdout(callback func(string)) {
 }
 
 func (p *Process) StreamToStdout() {
-	for len(p.outputBuffer) > 100 || time.Now().Sub(p.lastStream) > 100*time.Millisecond {
-		l := 100
+	//
+	// The output is buffered in the outputBuffer as it comes in from the TTY
+	// device.
+	//
+	// Ideally, we should strive to flush the output to the logfile as an event
+	// when there are enough data to be sent. "Enough data" in this context
+	// should satisfy the following criteria:
+	//
+	// - If there is more than 100 characters in the buffer
+	//
+	// - If there is less than 100 characters in the buffer, but they were in
+	//   the buffer for more than 100 milisecond. The reasoning here is that
+	//   it should take no more than 100 milliseconds for the TTY to flush its
+	//   output.
+	//
+	// - If the UTF-8 sequence is complete. Cutting the UTF-8 sequence in half
+	//   leads to undefined (?) characters in the UI.
+	//
+
+	maxTimeSinceLastFlush := 100 * time.Millisecond
+
+	for len(p.outputBuffer) > 100 || time.Now().Sub(p.lastStream) > maxTimeSinceLastFlush {
+		cutLength := 100
+
+		if len(p.outputBuffer) < cutLength {
+			cutLength = len(p.outputBuffer)
+		}
 
-		if len(p.outputBuffer) < l {
-			l = len(p.outputBuffer)
+		if time.Now().Sub(p.lastStream) < maxTimeSinceLastFlush {
+			//
+			// The unicode continuation sign is marked by the highest (8th) bit.
+			// If the bit is set, it means that the unicode character is not yet
+			// finished.
+			//
+			// In the bellow loop, we are cutting of the last 3 charactes in case
+			// they are marked as the unicode continuation characters.
+			//
+			// An unicode sequence can't be longer than 4 bytes
+			//
+			unicodeContinuationMask := uint(1 << 7)
```

Comment:
https://en.wikipedia.org/wiki/UTF-8
bug: continuation bytes start with `10xxxxxx`; `1xxxxxxx` also include leading bytes.

Reply:
This happens very frequently when doing box drawing
```
┏┳━━━━━━━━━━━━━���━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━���━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━���━━
┃┃ TEST Report
┗┻━━━━━━━━━━━━━━━━━━━━���━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━���━━━━━━━━━━━━━━━━━━━━━━━━━━━
```
Box drawing characters are 3 bytes.

Reply:
That's it! Thanks for finding the bug and sharing the solution. 🙇

Reply:
Work on the fix started here https://github.com/semaphoreci/agent/pull/79.

Reply:
@dnozay I started with the implementation of the patch, but unfortunately, I had a hard time replicating the issue in the Agent's tests.

My current assumption is that the root of the problem is not directly in the Agent, but upstream in the log processing service. We are investigating further.

I'll update as soon as we find something.